### ISSUE TITLE
关于Vit的pos embedding

### ISSUE THREAD

**Issue by Liqq1** [WZMIAOMIAO/deep-learning-for-image-processing#603]
您好！有两个问题想询问一下
1.讲解的时候说明，用1D位置编码 ，9个patch已经按照1-9 排好了；但在代码中为什么没有用到1-9这个位置顺序，需要学习才能得到？
<img width="785" alt="image" src="https://user-images.githubusercontent.com/54054113/180702963-d63a9c27-6237-4f92-b2dc-52c2a7f90ad0.png">
2.为什么class_token也需要pos_position,他不是固定好放在第0个位置吗？

谢谢

**Comment by WZMIAOMIAO** [comment]
1. 位置编码是学习得到的
2. 自己看下forward函数里是怎么concat拼接

**Comment by Liqq1** [comment]
> * 位置编码是学习得到的
> * 自己看下forward函数里是怎么concat拼接

2. forward里面是指定dim=0 去concat的，所以这个位置编码也需要学习吗，而不是直接固定为0？

**Comment by Liqq1** [comment]
> * 位置编码是学习得到的
> * 自己看下forward函数里是怎么concat拼接
这个是默认拼接到第0个位置上
 ` x = torch.cat((cls_token, x), dim=1)  # [B, 197, 768]`
包括下面的return，也是直接从位置0切片
` return self.pre_logits(x[:, 0])`
所以感觉class_token的位置固定是0，为什么还需要加上pos_positio学习得到(._.)？

**Comment by WZMIAOMIAO** [comment]
建议你还是再看下视频吧。

**Comment by Liqq1** [comment]
我又重新看了一下视频，视频里35min46s时说：“在cancat拼接的时候是将class_token放到最前面的，通过切片形式取最前面的数据。”
不知道我有没有理解错：我认为这样直接concat到前面，然后也是通过切片形式直接提取最前面的数据，所以class_token的位置固定是0，它也加上pos_position，只是为了保持维度一致，方便后续计算。

**Comment by WZMIAOMIAO** [comment]
是的

**Comment by Liqq1** [comment]
嗷嗷好的，谢谢
